### ISSUE TITLE
Refactor HWID system

### ISSUE THREAD

**Issue by Aciz** [etjump/etjump#1559]
The current HWID system doesn't really make sense, we calculate the HWID on client and mash everything together, and send a single hash to the server. This means that it's actually harmful to add new identification components to a HWID, because that means that in order to avoid e.g. a HWID ban, the user has more attack surface to spoof any given component, and a single successful component spoof will create a new HWID and avoid detection.

For the system to make more sense, and to make HWID ban evasion more difficult, the following should be implemented:

* Send each component of HWID individually on client, instead of a combined hash
* Store hash for each component of a HWID in user database, rather than a single combined hash
* Implement some sort of "scoring" system for each HWID component, based off the difficulty/effort required to change/spoof it
  * If a user successfully spoofs some component to circumvent a HWID ban detection, we check individual components and use this scoring system to detect potential ban evasions
  * Possible ban evasions can be logged to admin logs, and possibly sent to admin chat on server if it's enabled, to immediately notify any admins online on a server

**Comment by rudolfw8** [comment]
The system relies on the client to send HWID components. A determined attacker could modify the client to spoof all components before they are sent to the server.

**Comment by Aciz** [comment]
> The system relies on the client to send HWID components. A determined attacker could modify the client to spoof all components before they are sent to the server.

Yes of course, and that's how the current system works too. But there's no way around that, you have to request the client to send the HWID to the server. The main thing to improve here is that instead of a single combined hash for all components gobbled together, the server requests multiple hashes for each component and stores them individually, so a client has to spoof multiple components to appear as completely unique, instead of a single one.

I'm not sure if you have some idea on how to not get a HWID from client without the client actually calculating and sending it, I don't see how that would ever work.

**Comment by rudolfw8** [comment]
That’s true, but if the client can be modified, adding more components doesn’t really make it harder to spoof. An attacker could still spoof each individual component without changing the actual machine’s HWID. So, I’m curious—why add more components if they can just be easily modified on the client side to make it look unique, while the HWID remains unchanged?

**Comment by Aciz** [comment]
Because in a project like this, which is entirely open source down to the engine, it's not about making something entirely bulletproof that cannot be circumvented, that is never going to happen, it's about raising the effort required to do things like that so that most people will not bother.
